crates/ty_python_semantic/resources/mdtest/final.md

```diff
@@ -29,3 +29,344 @@ class H(
+
+class Foo:
+    @final
+    @identity
+    @identity
+    @identity
+    @identity
+    @identity
+    @identity
+    @identity
+    @identity
+    @identity
+    @identity
+    @identity
+    @identity
+    @identity
+    @identity
+    @identity
+    @identity
+    @identity
+    @identity
+    def bar(self): ...
+
+class Baz(Foo):
+    def bar(self): ...  # error: [override-of-final-method]
+```
+
+## Diagnostic edge case: superclass with `@final` method has the same name as the subclass
+
+<!-- snapshot-diagnostics -->
+
+`module1.py`:
+
+```py
+from typing import final
+
+class Foo:
+    @final
+    def f(self): ...
+```
+
+`module2.py`:
+
+```py
+import module1
+
+class Foo(module1.Foo):
+    def f(self): ...  # error: [override-of-final-method]
+```
+
+## Overloaded methods decorated with `@final`
+
+In a stub file, `@final` should be applied to the first overload. In a runtime file, `@final` should
+only be applied to the implementation function.
+
+<!-- snapshot-diagnostics -->
+
+`stub.pyi`:
+
+```pyi
+from typing import final, overload
+
+class Good:
+    @overload
+    @final
+    def bar(self, x: str) -> str: ...
+    @overload
+    def bar(self, x: int) -> int: ...
+    @final
+    @overload
+    def baz(self, x: str) -> str: ...
+    @overload
+    def baz(self, x: int) -> int: ...
+
+class ChildOfGood(Good):
+    @overload
+    def bar(self, x: str) -> str: ...
+    @overload
+    def bar(self, x: int) -> int: ...  # error: [override-of-final-method]
+    @overload
+    def baz(self, x: str) -> str: ...
+    @overload
+    def baz(self, x: int) -> int: ...  # error: [override-of-final-method]
+
+class Bad:
+    @overload
+    def bar(self, x: str) -> str: ...
+    @overload
+    @final
+    # error: [invalid-overload]
+    def bar(self, x: int) -> int: ...
+    @overload
+    def baz(self, x: str) -> str: ...
+    @final
+    @overload
+    # error: [invalid-overload]
+    def baz(self, x: int) -> int: ...
+
+class ChildOfBad(Bad):
+    @overload
+    def bar(self, x: str) -> str: ...
+    @overload
+    def bar(self, x: int) -> int: ...  # error: [override-of-final-method]
+    @overload
+    def baz(self, x: str) -> str: ...
+    @overload
+    def baz(self, x: int) -> int: ...  # error: [override-of-final-method]
+```
+
+`main.py`:
+
+```py
+from typing import overload, final
+
+class Good:
+    @overload
+    def f(self, x: str) -> str: ...
+    @overload
+    def f(self, x: int) -> int: ...
+    @final
+    def f(self, x: int | str) -> int | str:
+        return x
+
+class ChildOfGood(Good):
+    @overload
+    def f(self, x: str) -> str: ...
+    @overload
+    def f(self, x: int) -> int: ...
+    # error: [override-of-final-method]
+    def f(self, x: int | str) -> int | str:
+        return x
+
+class Bad:
+    @overload
+    @final
+    def f(self, x: str) -> str: ...
+    @overload
+    def f(self, x: int) -> int: ...
+    # error: [invalid-overload]
+    def f(self, x: int | str) -> int | str:
+        return x
```

Comment:
I find the diagnostic range here confusing. I read the message like 4 times and was confused why ty would think that the definition of `f` on line 248 isn't the implementation when it clearly is. I think we should change the diagnostic range to underline the `@final` on line 243 because that's where the error really is and is also the code you have to change (move it down to line 247)

Reply:
The diagnostic range for the `invalid-overload` diagnostic? Sure, but I'm not touching any of the `invalid-overload` logic in this PR :-) This test is just demonstrating how `override-of-final-method` diagnostics are emitted on methods that would _also_ be flagged as having invalid overloads: it's important to see how the two diagnostics interact